src/sources/aws_s3/sqs.rs

```diff
@@ -0,0 +1,543 @@
+
+                let mut send_error: Option<futures01::sync::mpsc::SendError<Event>> = None;
+                out.send_all(Compat::new(Box::pin(stream)))
+                    .compat()
+                    .await
+                    .map_err(|err| {
+                        send_error = Some(err);
+                        ()
+                    })
+                    .ok();
+
+                read_error
+                    .map(|error| {
+                        Err(ProcessingError::ReadObject {
+                            source: error,
+                            bucket: s3_event.s3.bucket.name.clone(),
+                            key: s3_event.s3.object.key.clone(),
+                        })
+                    })
+                    .unwrap_or(
+                        send_error
+                            .map(|error| {
+                                Err(ProcessingError::PipelineSend {
+                                    source: error,
+                                    bucket: s3_event.s3.bucket.name.clone(),
+                                    key: s3_event.s3.object.key.clone(),
+                                })
+                            })
+                            .unwrap_or(Ok(())),
+                    )
+            }
+            None => Ok(()),
+        }
+    }
+
+    async fn receive_messages(&self) -> Result<Vec<Message>, RusotoError<ReceiveMessageError>> {
+        self.sqs_client
+            .receive_message(ReceiveMessageRequest {
+                queue_url: self.queue_url.clone(),
+                max_number_of_messages: Some(10),
+                visibility_timeout: Some(self.visibility_timeout_secs),
+                ..Default::default()
+            })
+            .map_ok(|res| res.messages.unwrap_or_default())
+            .await
+    }
+
+    async fn delete_message(
+        &self,
+        receipt_handle: String,
+    ) -> Result<(), RusotoError<DeleteMessageError>> {
+        self.sqs_client
+            .delete_message(DeleteMessageRequest {
+                queue_url: self.queue_url.clone(),
+                receipt_handle,
+            })
+            .await
+    }
+}
+
+// https://docs.aws.amazon.com/AmazonS3/latest/dev/notification-content-structure.html
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[serde(rename_all = "PascalCase")]
+struct S3Event {
+    records: Vec<S3EventRecord>,
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+struct S3EventRecord {
+    event_version: semver::Version,
+    event_source: String,
+    aws_region: String,
+    event_name: S3EventName,
+
+    s3: S3Message,
+}
+
+#[derive(Clone, Debug)]
+struct S3EventName {
+    kind: String,
+    name: String,
+}
+
+// https://docs.aws.amazon.com/AmazonS3/latest/dev/NotificationHowTo.html#supported-notification-event-types
+//
+// we could enums here, but that seems overly brittle as deserialization would  break if they add
+// new event types or names
```

Comment:
Can't we still use enums, but have a catch all that at least logs if something is received that's unexpected (or just ignore "other")?

Still, this seems fine by me as well 👍

Reply:
:+1: I'll investigate this.

Reply:
I came back around the other way on this one :smile: I'd like to, at least, recognize any new `ObjectCreated` events without needing to add new enum kinds. I'll just leave this as-is until we find a need.